searx/query.py

```diff
@@ -63,100 +70,225 @@ def _parse_query(self):
+                        self.specific = True
+                elif enable_autocomplete:
+                    self._bang_autocomplete(query_part[0], value)
 
             # append query part to query_part list
-            if searx_query_part:
-                self.query_parts.append(query_part)
+            qlist = self.query_parts if searx_query_part else self.user_query_parts
+
+            qlist.append(query_part)
+            last_index_location = (qlist, len(qlist) - 1)
+
+        self.autocomplete_location = last_index_location
+
+    def _timeout_autocomplete(self):
+        self.autocomplete_list.append('<3')
+        self.autocomplete_list.append('<850')
+
+    def _timeout_parse(self, raw_value):
+        try:
+            raw_timeout_limit = int(raw_value)
+        except ValueError:
+            # error not reported to the user
+            return False
+        else:
+            if raw_timeout_limit < 100:
+                # below 100, the unit is the second ( <3 = 3 seconds timeout )
+                self.timeout_limit = float(raw_timeout_limit)
             else:
-                self.user_query_parts.append(query_part)
+                # 100 or above, the unit is the millisecond ( <850 = 850 milliseconds timeout )
+                self.timeout_limit = raw_timeout_limit / 1000.0
+            return True
+
+    def _language_autocomplete(self, value):
+        if len(value) == 0:
+            # show some example queries
+            self.autocomplete_list.append(":en")
+            self.autocomplete_list.append(":en_us")
+            self.autocomplete_list.append(":english")
+            self.autocomplete_list.append(":united_kingdom")
+        else:
+            for lc in language_codes:
+                lang_id, lang_name, country, english_name = map(str.lower, lc)
+
+                # check if query starts with language-id
+                if lang_id.startswith(value):
+                    if len(value) <= 2:
+                        self.autocomplete_list.append(':' + lang_id.split('-')[0])
+                    else:
+                        self.autocomplete_list.append(':' + lang_id)
+
+                # check if query starts with language name
+                if lang_name.startswith(value) or english_name.startswith(value):
+                    self.autocomplete_list.append(':' + lang_name)
+
+                # check if query starts with country
+                if country.startswith(value.replace('_', ' ')):
+                    self.autocomplete_list.append(':' + country.replace(' ', '_'))
+
+    def _language_parse(self, value, enable_autocomplete):
+        found = False
+        # check if any language-code is equal with
+        # declared language-codes
+        for lc in language_codes:
+            lang_id, lang_name, country, english_name = map(str.lower, lc)
+
+            # if correct language-code is found
+            # set it as new search-language
+
+            if (value == lang_id
```

Comment:
Please extract this long condition list into a helper function with a descriptive name. E.g. `is_valid_language_code`.

Reply:
It can be 
```python
    def match_language(lang_id, lang_name, english_name, country):
       return value == lang_id
                  or value == lang_name
                  or value == english_name
                  or value.replace('-', ' ') == country

    def _parse(self, value):
        found = False
        # check if any language-code is equal with
        # declared language-codes
        for lc in language_codes:
            lang_id, lang_name, country, english_name = map(str.lower, lc)

            # if correct language-code is found
            # set it as new search-language

            if match_language(lang_id, lang_name, english_name, country) \
               and value not in self.raw_text_query.languages:
                found = True
                ...
```

Related to #2240

Can it be the topic of another PR ?

Reply:
Sure.